Rakefile

```diff
@@ -1,66 +1,116 @@
-require "json"
 require "pathname"
+require "time"
 
-PROJ_DIR     = Pathname.new(__dir__)
-CONFIG_FILE  = PROJ_DIR.join(ENV["CNFFILE"] || "config.json")
-BUILD_ROOT   = Pathname.new(ENV["OBJDIR"] || "build")
-INSTALL_ROOT = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+PROJ_DIR          = Pathname.new(__dir__)
+APPS_DIR          = PROJ_DIR.join('apps')
+GEMFILE           = PROJ_DIR.join('Gemfile')
+INSTALL_ROOT      = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+VENDOR_BUNDLE     = (ENV['VENDOR_BUNDLE'] == "yes" || ENV['VENDOR_BUNDLE'] == "true")
+PASSENGER_APP_ENV = ENV["PASSENGER_APP_ENV"] || "production"
 
-def all_components
-  JSON.parse(CONFIG_FILE.read).map { |c| Component.new(c) }
+def apps
+  Dir["#{APPS_DIR}/*"].map { |d| Component.new(d) }
 end
 
 class Component
   attr_reader :name
-  attr_reader :url
-  attr_reader :tag
-
-  def initialize(opts = {})
-    @name = opts.fetch("name") { raise ArgumentError, "No name specified. Missing argument: name" }.to_s
-    @url  = opts.fetch("url")  { raise ArgumentError, "No url specified. Missing argument: url" }.to_s
-    @tag  = opts.fetch("tag")  { raise ArgumentError, "No tag specified. Missing argument: tag" }.to_s
-    @app  = opts.fetch("app", nil)
+  attr_reader :path
+
+  def initialize(app)
+    @name = File.basename(app)
+    @path = Pathname.new(app)
   end
 
-  def app?
-    !!@app
+  def ruby_app?
+    @path.join('config.ru').exist?
   end
 
-  def build_root
-    app? ? BUILD_ROOT.join("apps", name) : BUILD_ROOT.join(name)
+  def node_app?
+    @path.join('app.js').exist?
   end
+end
 
-  def download_url
-    "#{url}/archive/#{tag}.tar.gz"
+desc "Package OnDemand"
+task :package do
+  `which gtar 1>/dev/null 2>&1`
+  if $?.success?
+    tar = 'gtar'
+  else
+    tar = 'tar'
   end
+  version = ENV['VERSION']
+  if ! version
+    latest_commit = `git rev-list --tags --max-count=1`.strip[0..6]
+    latest_tag = `git describe --tags #{latest_commit}`.strip[1..-1]
+    datetime = Time.now.strftime("%Y%m%d-%H%M")
+    version = "#{latest_tag}-#{datetime}-#{latest_commit}"
+  end
+  sh "git ls-files | #{tar} -c --transform 's,^,ondemand-#{version}/,' -T - | gzip > packaging/v#{version}.tar.gz"
 end
 
-task :default => :build
+namespace :build do
+  desc "Build gems"
+  task :gems do
+    if VENDOR_BUNDLE
+      bundle_args = "--path vendor/bundle"
```

Comment:
We don't this `VENDOR_BUNDLE` flag.

If I want to bundle install all of the apps to a relative path like vendor/bundle I can just do this:

    BUNDLE_PATH=vendor/bundle rake build:gems

That will use the path relative to each app directory - so separate vendor/bundles. If I want them to share a path I can do this:

    BUNDLE_PATH=/Users/efranz/dev/ondemand/shared/vendor/bundle rake build:gems

The absolute path does the trick here. Of course it might be better again to do a relative from the app directory, especially if we are moving/copying the app directories. But this is just for an example to show I can really control the location using Bundler's support for config via environment variables.

Reply:
well... maybe I take that back. Using BUNDLE_PATH lets you control where gems are installed, but the setting is not preserved in the app's .bundle/config like the command line arguments are.